lib/cog/template.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Cog.Template do
+  alias Cog.Queries
+  alias Cog.Repo
+  alias Cog.TemplateCache
+
+  @fallback_adapter "any"
+
+  def render(adapter, bundle_id, template, context) do
+    with {:ok, template_fun} <- fetch_compiled_fun(adapter, bundle_id, template, context) do
+      template_fun.(context)
+    end
+  end
+
+  def fetch_compiled_fun(adapter, bundle_id, template, context) do
+    with :error              <- TemplateCache.lookup(adapter, bundle_id, template),
+         {:ok, source}       <- fetch_source(adapter, bundle_id, template, context),
+         {:ok, template_fun} <- compile(source),
+         :ok                 <- TemplateCache.insert(adapter, bundle_id, template, template_fun),
+         do: {:ok, template_fun}
```

Comment:
Just some documentation explaining the rationale.

I guess the only thing that I'm not super jazzed about with these "negative `with`" calls is that, absent some specs or docs, it's a little hard to figure out what all you can expect to return (though, to be fair, I think that applies equally to "positive `with`" calls, as well).